### DISCUSSION TITLE
How to get the name of calling feature file from step definition?

### DISCUSSION THREAD

**Discussion by sachingainewar** [cypress-io/cypress#26024] [Questions and Help]
My framework contain `common step definitions `which calls from different `feature `files. I want to set the `locators `base on the `step definition` calls from feature file.

For example below `Steps `calling `step Definition` from common module.

```
A1.feature
Then I click on "Save" button

B2.feature
Then I click on "Save" button

```
Common step definition

```
const A1= require('../../../support/locators/A1-loc.cy')
const B2= require('../../../support/locators/B2-loc.cy')


Then ('I click on {string} button', (buttonName)=>{
    clickOnButton(getButtonLocator(fromButtonNameAndFeatureFileName)
})
```
I want to choose the different locators for `save `button base on different feature file. Means if `A1.feature` file calls then point to `A1-loc.cy` and if` B2.feature` file calls then points to` B2-loc.cy`

**Comment by portyu9** [comment]
Hi @sachingainewar 

With the current `@badeball/cypress-cucumber-preprocessor`, you can get the feature URI from the Cucumber hook arguments and make it available to the shared step.

For example:

```ts
import {
  Before,
  Then,
} from '@badeball/cypress-cucumber-preprocessor'

let featureUri = ''

Before(function ({ pickle }) {
  featureUri = pickle.uri
})

Then('I click on {string} button', (buttonName) => {
  const featureName = featureUri.split(/[\\/]/).pop()

  // use featureName/buttonName to choose the locator
})
```

`pickle.uri` identifies the source feature file.

That said, if `A1.feature` and `B2.feature` require different locator implementations for the same natural-language step, I would prefer **feature-specific step definition files** rather than branching on the filename inside one global step.

The maintained preprocessor supports configuring `stepDefinitions` with `[filepath]` / `[filepart]` patterns so feature-local steps can live alongside the relevant feature while genuinely common steps remain shared.

Also note that the old unscoped `cypress-cucumber-preprocessor` package is obsolete; for a current Cypress project use `@badeball/cypress-cucumber-preprocessor`.

Please mark this answer as accepted if it helped, thank you.
